src/compiler/checker.ts

```diff
@@ -28839,7 +28837,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     // type of T.
     function getContextualTypeForElementExpression(arrayContextualType: Type | undefined, index: number): Type | undefined {
         return arrayContextualType && (
-            getTypeOfPropertyOfContextualType(arrayContextualType, "" + index as __String)
+            getTypeOfPropertyOfContextualType(filterType(arrayContextualType, t => !!getIndexTypeOfType(t, numberType) || isAssignableToAvailableAnyIterable(t)), "" + index as __String)

```

Comment:
The `!!getIndexTypeOfType(t, numberType)` bit here is required to cover [this `concatTuples` test case](https://github.com/microsoft/TypeScript/blob/ed15865eb065006da26a233d07c7899103b67c08/tests/cases/compiler/concatTuples.ts). `.concat` is typed using `ConcatArray` which isn't exactly an array.